src/core/beforeSave.js

```diff
@@ -0,0 +1,40 @@
+import { docLink, showError } from "./utils.js";
+import { sub } from "./pubsubhub.js";
+
+export const name = "core/beforeSave";
+
+export function run(conf) {
+  if (!conf.beforeSave) return;
+
+  if (
+    !Array.isArray(conf.beforeSave) ||
+    conf.beforeSave.some(
+      el => typeof el !== "function" || el.constructor.name === "AsyncFunction"
+    )
+  ) {
+    const msg = docLink`${"[beforeSave]"} configuration option must be an array of synchronous JS functions.`;
+    showError(msg, name);
+    return;
+  }
+
+  sub(
+    "beforesave",
+    ({ ownerDocument }) => {
+      let pos = 0;
+      for (const fn of conf.beforeSave) {
+        try {
+          fn(ownerDocument);
```

Comment:
Should it also have access to `conf`?

Reply:
I don't think so, tbh. The conf is already global and we want to discourage people poking at the private stuff.